[FLINK-1502] [WIP] Basic Metric System

zentol opened the pull request:
This PR is a preview of the new metric system. 
It is not complete because
* there is no documentation for the website
* a few smaller parts also don't have code documentation
* I haven't tried out the ganglia/statsD reporter yet

In general though it works and it is now time to gather some feedback.

The PR is organized into several commits to give it some structure; generally divided by which part of the system they expose the metric system to. Note that  The last commit "Metric Usage Examples" is not technically part of the PR but showcases the usage.
The division was done very simple, so some changes may technically belong to several commits.
## General overview

A user can access a system-provided MetricGroup to register a Metric, which is stored in a MetricRegistry and forwarded regularly to a Reporter which communicates them to an external system.

## MetricGroups

MetricGroups are the user-facing part of the system. They are a nested data structure, containing other groups and metrics, that allow registering metrics with Flink while organizing them in a hierarchy.

For example, every TaskManager has a MetricGroup, and for every task that is deployed a new sub-group for that task is added. This task specific group is propagated through the task stack, with new groups/metrics being added. Within a UDF the operator MetricGroup is accessed through the RuntimeContext.

## Metrics

Metrics are the objects used to measure something.

Metrics include 
* Gauges, that measure a value on-demand
* Meters, that measure the rate/count of events
* Histograms, that measure the distribution of long values
* Counters, that count stuff
* Timers, that measure rate of calls and distribution of execution time for a given piece of code.

Under the hood we use the Metrics from the Dropwizard library. In order to ensure interface stability, and to give us the option to reimplement things without breaking everything, they (and other classes) are wrapped to match our interfaces. 

## Reporters

Reporters are the component that communicate the Metrics to the outside world. With this PR we allow exporting Metrics via JMX (default), Graphite, Ganglia and StatsD. They interval in which they report is configurable.

Similarly to Metrics, we partially use reporters from the DropWizard library (Graphite, Ganglia), again wrapped to match out interfaces.

Reporters are configured via flink-conf.yaml.

An example configuration might look like this:
metrics.reporter.class: org.apache.flink.metrics.GraphiteReporter
metrics.reporter.arguments: --host localhost --port 8080
metrics.reporter.interval: 30 SECONDS

Reporters are instantiated generically and configured with a Configuration containing the parsed arguments. All non-JMXReporters are not part of the distribution and have to be added to the classpath manually (usually by putting the jar into /lib)

JMX uses the port 9010 by default, This can be configured by setting the metrics.jmx.port property in the flink-conf.yaml

## Registry

The registry is essentially just a connection between all MetricGroups and the Reporter.

## System Scope

The system scope for a metric (from which the metric name is generated) can be modified by setting a format string using the metrics.scope.\<component> key in the flink-conf.yaml. Valid components are “taskmanager”, “job”, “task”, “operator”. This will affect all metrics that are directly tied to that component, without affecting parent-/subcomponents.

Format strings may contain constants (e.g “task”) and variables (e.g “\<task_id>”) which will be replaced at runtime. 

Variables may refer to a property of the component or any parent components. For example, by setting the following setting the scope for all metrics that are tied to a task would be “WordCount.task.MyTask”.
 
metrics.scope.task: \<job_name>.task.\<task_name>

There is no restrictions on the order of properties.

If the user simply wants to use the default scope for parent components he can prepend his format string with ‘*’.

Full reference for all properties:

TaskManager: \<host>, \<tm_id>
Job: \<job_id>, \<job_name>
Task: \<task_id>, \<task_name>, \<task_attempt>, \<subtask_index>
Operator: \<operator_name>, \<subtask_index>

Default format strings for each component:
TaskManager: \<host>.taskmanager.\<tm_id>
Job: \<host>.taskmanager.\<tm_id>.\<job_name>
Task: \<host>.taskmanager.\<tm_id>.\<job_name>.\<task_name>.\<subtask_index>
Operator: \<host>.taskmanager.\<tm_id>.\<job_name>.\<operator_name>.\<subtask_index>

rmetzger commented on flink-metrics/flink-metrics-ganglia/pom.xml:
```diff
@@ -0,0 +1,84 @@
+		<dependency>
+			<groupId>info.ganglia.gmetric4j</groupId>
+			<artifactId>gmetric4j</artifactId>
+			<version>1.0.7</version>
```
License seems to be ok: https://github.com/ganglia/gmetric4j/blob/master/COPYING

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -683,6 +688,9 @@ else if (STATE_UPDATER.compareAndSet(this, current, ExecutionState.FAILED)) {
 				libraryCache.unregisterTask(jobId, executionId);
+				
+				//Uncomment before Merging!!!
+				//metrics.close();
```
why not now?

zentol replied:
it makes it a bit easier to try out; without this the metrics are no longer exposed when the job is finished. so if you run one of the examples with the built-in data you'll see nothing (probably).

rmetzger replied:
Okay, I see.

uce commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -152,6 +154,9 @@ class TaskManager(
   /** Registry of metrics periodically transmitted to the JobManager */
   private val metricRegistry = TaskManager.createMetricsRegistry()
```
What are the plans for removing/subsuming this?

zentol replied:
when we can display the new metrics in the dashboard.

uce commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -924,6 +936,15 @@ class TaskManager(
       libraryCacheManager = Some(new FallbackLibraryCacheManager)
     }
+
+    metricsRegistry = new FlinkMetricRegistry(GlobalConfiguration.getConfiguration);
```
I think it's better to use `config.configuration`. At some point in time we might get around to removing the `GlobalConfiguration`.

rmetzger commented:
Is it possible that submitting the same job two times doesn't work?

```
./bin/flink run ./examples/streaming/SocketWindowWordCount.jar --port 54323

05/05/2016 19:09:08	Job execution switched to status RUNNING.
05/05/2016 19:09:08	Source: Socket Stream -> Flat Map(1/1) switched to SCHEDULED 
05/05/2016 19:09:08	Source: Socket Stream -> Flat Map(1/1) switched to DEPLOYING 
05/05/2016 19:09:08	Fast SlidingProcessingTimeWindows(5000, 1000) of WindowedStream.main(SocketWindowWordCount.java:79) -> Sink: Unnamed(1/1) switched to SCHEDULED 
05/05/2016 19:09:08	Fast SlidingProcessingTimeWindows(5000, 1000) of WindowedStream.main(SocketWindowWordCount.java:79) -> Sink: Unnamed(1/1) switched to DEPLOYING 
05/05/2016 19:09:08	Source: Socket Stream -> Flat Map(1/1) switched to RUNNING 
05/05/2016 19:09:08	Source: Socket Stream -> Flat Map(1/1) switched to FAILED 
java.lang.IllegalArgumentException: This group ([key0, localhost, Actor, TaskManager, TaskManager, 964566cfcf032710aff86614010fce21, Category, Tasks, Job, "Socket Window WordCount", Operator, Flat Map, SubTask, 0, ChannelType, OutputChannel, Index, 0]) already contains a metric named numBytesOut
	at org.apache.flink.metrics.MetricGroup.addMetric(MetricGroup.java:246)
	at org.apache.flink.metrics.MetricGroup.counter(MetricGroup.java:123)
	at org.apache.flink.runtime.io.network.api.serialization.SpanningRecordSerializer.setMetrics(SpanningRecordSerializer.java:206)
	at org.apache.flink.runtime.io.network.api.writer.RecordWriter.setMetrics(RecordWriter.java:216)
	at org.apache.flink.streaming.runtime.tasks.OperatorChain.createStreamOutput(OperatorChain.java:294)
	at org.apache.flink.streaming.runtime.tasks.OperatorChain.<init>(OperatorChain.java:94)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:188)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:584)
	at java.lang.Thread.run(Thread.java:745)

05/05/2016 19:09:08	Fast SlidingProcessingTimeWindows(5000, 1000) of WindowedStream.main(SocketWindowWordCount.java:79) -> Sink: Unnamed(1/1) switched to RUNNING 
05/05/2016 19:09:08	Job execution switched to status FAILING.

```

rmetzger commented:
I'm just assuming its the missing "close()" call's I've commented already causing this issue.

I just tried adding some custom metrics:
```java
new RichFlatMapFunction<String, WordWithCount>() {
					public Counter output;
					public Counter el;
					public String lastOut;

					@Override
					public void open(Configuration parameters) throws Exception {
						super.open(parameters);
						MetricGroup mg = getRuntimeContext().getMetricGroup();

						this.el = mg.counter("elements");

						MetricGroup detailedGroup = mg.addGroup("detailed");
						this.output = detailedGroup.counter("output");
						detailedGroup.gauge("lastOut", new Gauge<String>() {
							@Override
							public String getValue() {
								return lastOut;
							}
						});
					}

					@Override
					public void flatMap(String value, Collector<WordWithCount> out) {
						el.inc();
						for (String word : value.split("\\s")) {
							lastOut  = word;
							out.collect(new WordWithCount(word, 1L));
							output.inc();
						}
					}
				}
```

and it works amazingly well

![image](https://cloud.githubusercontent.com/assets/89049/15050706/9ecc6fda-12f5-11e6-8c7d-cc7cb3657ecd.png)

ankitcha commented:
Guys, Thanks you for this awesome addition to Flink. 

I was just wondering if there is a way to configure multiple reporters as well?

zentol commented:
@rmetzger you are correct that you're job failed since the previous one wasn't cleaned up yet. Should you try to run 2 identical jobs in parallel it will fail, since 2 jobs would use the same metrics due to name clashes. Note that in this version this also occurs when 2 operators have the same name. I have some additional functionality coming up that would allow you to circumvent this issue.

@ankitcha The problem with multiple reporters is our configuration, it only supports single-line key-value pairs, and you need to know the exact key to access it. In order to configure multiple reporters you would either need a nested structure (which is not supportet), or index the configuration keys (metrics.reporter.1.class) and add a new parameter containing the indices to use (e.g. metrics.reporter: 0, 1), which isn't particularly user-friendly. The metric system itself could deal with multiple reporters with minor modifications; it's all about the configuration.

rmetzger commented:
Okay. I think we need to support multiple instances of the same job on a TaskManager.

ankitcha commented:
@rmetzger thanks for the response. I think I got what you explained and I agree that using index in configuration keys won't be a nice experience. 

But, can we support nested structure in flink conf? I am unsure about the scope of this change, so maybe its a bad suggestion. But, this is something that will really help me out to put our application in production and we have to use multiple reporters as part of our infrastructure requirements.

zentol commented:
@ankitcha you could write your own reporter that simply wraps the reporters you want to use, simply forwarding the calls to all of them. this should actually be really easy.

ankitcha commented:
@zentol thanks for the suggestion. I think that can work, I will try that and confirm.

Thanks!

zentol commented:
I have updates this PR; it now contains a new feature: configurable system scope. It essentially allows users to decide how Flink entities(Taskmanager, jobs, tasks and operators) are represented in metric names.

Users can configure a different format strings for every entity type in the flink-conf.yaml; which format is applied to a metric depends to which entity it is bound to.

For example, users can do the following:
* omit taskmanager information for jobs/tasks/operators
* re-order properties, like having the job properties ahead of the taskmanager
* decided whether they want to use names, ID's or a mix of both!

I've updated the main post to include more information.

StephanEwen commented:
Looks nice, I am trying it out now

zentol commented:
I'm currently aware of the following issues:
* quite a few tests in flink-runtime fail; this is mostly due to a missing integration into the MockEnvironments. Fix coming up
* there is an issue in the IO metrics for operations with multiple outputs. This one is a bit more tricky, but a not-so-pretty solution is on the way.

zentol commented:
The tests should now pass.

I've fixed the following issues:
* several examples had non-transient Metric fields
* removed one failing example (CollectionInputFormat; accessed Context in constructor)
* IOMetrics are not granular, effectively gathering the same metrics we gather currently
* Resolved several test issues
 * General:
   * Introduced several DummyMetricGroups for use in tests
    * mocked Context/Environments now return DummyMetricGroups
 * Core:
   * RuntimeUDFContext constructor modified to take a MetricGroup argument
    * Integrated metrics into the CollectionExecutor
 * Table API: asterisks (introduced by select *) were not properly removed by the JMXReporter
 * Streaming: StreamingConfig ChainIndex default is now 0 instead of -1

StephanEwen commented:
This looks pretty good now.
I would like to get this in soon, now that the test are passing.
Let's iterate over it on the master.

This needs a rebase to master, though,

Also saw that you have a "metrics_v4" branch now. Is that one newer than the pull request?

zentol commented:
The code in this PR is the most up-to-date.

StephanEwen commented:
I rebased the branch, waiting for Travis to give a green light...

StephanEwen commented:
Tests were good!
Rebased again, re-running tests, will merge after that.

StephanEwen commented:
Hmm, seeing some local test failures after the re-base. Have to look into this.
Typical message is `java.lang.IllegalArgumentException: This group already contains a metric named KeyCount`

StephanEwen commented:
Few followups are actually needed before merging this:

  1. We need to remove the example metrics

  2. Conflicting metrics names should not result in failures. Metrics are tooling, and problems in the tooling should not fail the core programs.

  3. I think we should limit the available metrics types to Gauge and Counter for now. I looked at Timers, Meters, and Histograms - they are very high overhead each. As a follow-up, I would like to see if we can construct simple Meters as views over counters. That way, the runtime code as no overhead for the metering (it just maintains counters and gauges) and the registry code needs to turn them into Meters asynchronously.

StephanEwen commented:
I would like to take an iteration on this, make some changes on top of this, and open a new pull request afterwards.

zentol commented:
I've had an offline chat with @StephanEwen and as a result updated the PR.

List of changes:
* dropped Meters, Histograms and Timers
 * we weren't convinced that the performance of the implementation used is efficient enough for our use-cases
 * removed Reservoir/Snapshot and various wrapper classes
* moved ScheduledDropWizardReporter into a new flink-metrics-dropwizard module within flink-metrics
 * removes Dropwizard usage entirely from flink-core
* the MetricRegistry no longer maintains maps of all metrics; reporters are from now on responsible for doing this
 * Listener interface was removed
 * AbstractReporter class was added that implements this behaviour
 * will make it easier to support multiple reporters in the future
* Counter and Gauge no longer implement the DropWizard interface
 * added Counter-/GaugeWrapper classes for DropWizard reporters

StephanEwen commented:
Thanks, Chesnay. Checking out the new changes now...

StephanEwen commented:
How about renaming the maven module from `flink-metrics` to `flink-metric-reporters`? The metrics systems itself is part of `flink-core` after all...

zentol commented:
renaming the module is a good idea

StephanEwen commented:
Let me grab the token for this one. There are a few things still, like resource leaks, in this code.
I'll pass you back the token as soon as I am done...

StephanEwen commented:
I did a pass over the code and committed the result:

Manually merged in 003ce18efc0249fae874e56c3df6acf19f5f2429 and 707606ac40dbbbd497fcbbb5442870fec5468bf3

sumitchawla commented:
@zentol is there any HTTP interface to these metrics that can be used to query the metrics? Something similar to existing JobManager Accumulators urls?

zentol commented:
No. They are currently only accessible via JMX or a system supported by a reporter. They are not available in the Dashboard, this will be worked on for 1.2 .

You could write a reporter that listens for http requests and returns the appropriate value.

sumitchawla commented:
hi @zentol .. thanks for the information.  This is great and i could see the metrics in JMX.  I have one more question on interoperability of Accumulators and Metrics.  As per my understanding, currently Metrics are only available at system level , and User Accumulators are available to Job writers.  Is there any plan for supporting Metrics for Job writers? Metrics give much more capabilities than current Accumulators and would be a great way to track custom metrics at each operator level?

zentol commented:
What do you mean by "Job writers"?

sumitchawla commented:
@zentol .. by job writer i meant end user writing jobs using Flink API.  As of now i can create custom accumulators using `getRuntimeContext().addAccumulator(ACCUMULATOR_NAME,...` can i do something similar to register custom metrics in my transformations

zentol commented:
@sumitchawla sure you can, as described here: https://ci.apache.org/projects/flink/flink-docs-master/apis/metrics.html#registering-metrics

sumitchawla commented:
Thanks a lot @zentol .. this is great..
